File: in_Java/0802  Find Eventual Safe States.java
class Solution {
    public static final int UNKNOWN = 0;  // unknown.
    public static final int UNSAFE = 1;   // unsafe
    public static final int SAFE   = 2;   // safe
    
    public List<Integer> eventualSafeNodes(int[][] graph) {
        List<Integer> res = new ArrayList<>();
        int[] states = new int[graph.length];
        for(int i = 0; i < graph.length; i++){
            if(dfs(graph, i, states)) res.add(i);
        }
        return res;
    }
    private boolean dfs(int[][] g, int i, int[] states){
        if(states[i] != UNKNOWN) return states[i] == SAFE;
        states[i] = UNSAFE;
        for(int nei : g[i]){
            if(!dfs(g,nei,states)) return false;
        }
        states[i] = SAFE;
        return true;
    }
}

File: in_Python/0802 Find Eventual Safe States.py
class Solution(object):
    def eventualSafeNodes(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[int]
        """
        def dfs(i):
            visited[i] = 0
            for ii in graph[i]:
                if visited[ii] == 0 or (visited[ii] == -1 and dfs(ii)):
                    return True
            res.append(i)
            visited[i] = 1
            return False
        
        visited = [-1] * len(graph)
        res = []
        for i in range(len(graph)):
            if visited[i] == -1:
                dfs(i)
        res.sort()
        return res